ClassFileParser

boredhuman opened the pull request:
ClassFileParser basically done

Commit:
start on classfileparser

Commit:
removed git issue and did more work to classfileparser

x4e commented on src/main/java/dev/binclub/javaception/classfile/ClassFileParser.java:
```diff
@@ -1,4 +1,235 @@
+	public ClassFileParser(byte[] classBytes) throws Throwable {
+		if (classBytes.length < 4) {
+			// need atleast 4 bytes get magic number
+			throw new Throwable("Invalid class file");
```
This should throw a `ClassFormatError` ([jvms 5.3.5](https://docs.oracle.com/javase/specs/jvms/se16/html/jvms-5.html#jvms-5.3.5)).

x4e commented on src/main/java/dev/binclub/javaception/classfile/ClassFileParser.java:
```diff
@@ -1,4 +1,235 @@
+		dis = new DataInputStream(new ByteArrayInputStream(classBytes));
+		int magic = dis.readInt();
+		if (magic != 0xCAFEBABE) {
+			throw new Throwable("Invalid Class File");
```
`ClassFormatError`

x4e commented on src/main/java/dev/binclub/javaception/classfile/ClassFileParser.java:
```diff
@@ -1,4 +1,235 @@
+		constantPoolCount = dis.readUnsignedShort();
+		constantPool = new Object[constantPoolCount - 1];
+		// force enum to resolve
+		System.out.println(ConstantTypes.values().length);
```
Perhaps just `var _ = ConstantTypes.values();`?

x4e commented on src/main/java/dev/binclub/javaception/classfile/ClassFileParser.java:
```diff
@@ -1,4 +1,235 @@
+		default:
+			System.out.println("skipping attrib type " + name);
+			for (int i = 0; i < attributeLength; i++) {
+				dis.readUnsignedByte();
```
I believe there is a `skipBytes` method that might be better to use

x4e commented on src/main/java/dev/binclub/javaception/classfile/ClassFileParser.java:
```diff
@@ -1,4 +1,235 @@
+		if (attributesCount != 0) {
+			attributes = new AttributeInfo[attributesCount];
+			for (int i = 0; i < attributesCount; i++) {
+				attributes[i] = readAttribute(dis, constantPool);
```
`readAttribute` can sometimes return null, meaning there will be unused array slots of null values, perhaps it would be more efficient to use an `ArrayList` and add attributes to it only if they are non null?

x4e commented on src/main/java/dev/binclub/javaception/classfile/ClassFileParser.java:
```diff
@@ -1,4 +1,235 @@
+			if (attributesCount != 0) {
+				attributes = new AttributeInfo[attributesCount];
+				for (int j = 0; j < attributesCount; j++) {
+					attributes[j] = readAttribute(dis, constantPool);
```
`readAttribute` can sometimes return null, meaning there will be unused array slots of null values, perhaps it would be more efficient to use an `ArrayList` and add attributes to it only if they are non null?

x4e commented on src/main/java/dev/binclub/javaception/classfile/ClassFileParser.java:
```diff
@@ -1,4 +1,235 @@
+			if (attributesCount != 0) {
+				attributes = new AttributeInfo[attributesCount];
+				for (int j = 0; j < attributesCount; j++) {
+					attributes[j] = readAttribute(dis, constantPool);
```
`readAttribute` can sometimes return null, meaning there will be unused array slots of null values, perhaps it would be more efficient to use an `ArrayList` and add attributes to it only if they are non null?

x4e commented on src/main/java/dev/binclub/javaception/classfile/ClassFileParser.java:
```diff
@@ -1,4 +1,235 @@
+			int nameIndex = dis.readUnsignedShort();
+			int descriptorIndex = dis.readUnsignedShort();
+			int attributesCount = dis.readUnsignedShort();
+			AttributeInfo[] attributes = null;
```
Would it not be better to have an empty attribute array by default?

x4e commented on src/main/java/dev/binclub/javaception/classfile/ClassFileParser.java:
```diff
@@ -1,4 +1,235 @@
+			int nameIndex = dis.readUnsignedShort();
+			int descriptorIndex = dis.readUnsignedShort();
+			int attributesCount = dis.readUnsignedShort();
+			AttributeInfo[] attributes = null;
```
Would it not be better to have an empty attribute array by default?

x4e reviewed (changes requested):
Looks great, thanks! Just needs a few changes

Commit:
ClassFileParser

all constant pool constants excluding records and module can be parsed
additionally bootstrap attribute can be parsed
still need to convert null containing arrays to arraylists

Commit:
Attribute arrays containing nulls now changed to lists

boredhuman commented:
All changes requested have been done
